cub/cub/util_device.cuh

```diff
@@ -69,6 +70,161 @@ CUB_NAMESPACE_BEGIN
+  // thread block but would still suffice for fallback agent's shared memory requirements
+  static constexpr bool uses_fallback_policy =
+    (sizeof(typename DefaultAgentT::TempStorage) > max_smem_per_block)
+    && (sizeof(typename FallbackAgentT::TempStorage) <= max_smem_per_block);
+
+public:
+  // The tuning policy and agent that will be used during algorithm invocation
+  using agent_t        = cub::detail::conditional_t<uses_fallback_policy, FallbackAgentT, DefaultAgentT>;
+  using agent_policy_t = cub::detail::conditional_t<uses_fallback_policy, FallbackAgentPolicyT, DefaultAgentPolicyT>;
+
+private:
+  // The amount of shared memory or virtual shared memory required by the algorithm's agent
+  static constexpr std::size_t required_shmem = sizeof(typename agent_t::TempStorage);
+
+  // Whether we need to allocate global memory-backed virtual shared memory
+  static constexpr bool needs_vsmem = required_shmem > max_smem_per_block;
+
+public:
+  // Type alias to be used for static temporary storage declaration within the algorithm's kernel
+  using static_temp_storage_t = cub::detail::conditional_t<needs_vsmem, cub::NullType, typename agent_t::TempStorage>;
+
+  // The amount of global memory-backed virtual shared memory needed
+  static constexpr std::size_t vsmem_per_block = needs_vsmem ? required_shmem : 0;
+
+  /**
+   * @brief Used from within the device algorithm's kernel to get the temporary storage that can be
+   * passed to the agent, specialized for the case when we can use native shared memory as temporary
+   * storage.
+   */
+  static __device__ __forceinline__ typename agent_t::TempStorage&
+  get_temp_storage(typename agent_t::TempStorage& static_temp_storage, vsmem_t&)
+  {
+    return static_temp_storage;
+  }
+
+  /**
+   * @brief Used from within the device algorithm's kernel to get the temporary storage that can be
+   * passed to the agent, specialized for the case when we have to use global memory-backed
+   * virtual shared memory as temporary storage.
+   */
+  static __device__ __forceinline__ typename agent_t::TempStorage&
+  get_temp_storage(cub::NullType& static_temp_storage, vsmem_t& vsmem)
+  {
+    return *reinterpret_cast<typename agent_t::TempStorage*>(
+      static_cast<char*>(vsmem.gmem_ptr) + (vsmem_per_block * blockIdx.x));
+  }
+
+  /**
+   * @brief Overload for the case when native shared memory can be used. No need to hint to discard
+   * modified cache lines.•
+   */
+  template <bool needs_vsmem_ = needs_vsmem, typename ::cuda::std::enable_if<!needs_vsmem_, int>::type = 0>
+  static __device__ __forceinline__ void discard_temp_storage(typename agent_t::TempStorage& temp_storage)
+  {}
+
+  /**
+   * @brief Overload for the case when virtual shared memory is used. Hints to discard modified
+   * cache lines for vshmem.
+   *
+   * @note Needs to be followed by `__syncthreads()` if the virtual shared memory is reused within the same thread
```

Comment:
Should this function return true/false to indicate to the caller whether or not the synchronization is needed? E.g. this overload return true, the non-vsmem overload returns false.

Reply:
That's a good point. I'll add it 👍